rust/lance/src/dataset/cleanup.rs

```diff
@@ -193,50 +210,92 @@ impl<'a> CleanupTask<'a> {
         Ok(())
     }
 
-    async fn delete_unreferenced_files(&self) -> Result<RemovalStats> {
+    async fn delete_unreferenced_files(
+        &self,
+        inspection: CleanupInspection,
+    ) -> Result<RemovalStats> {
         let removal_stats = Mutex::new(RemovalStats::default());
+        let verification_threshold = utc_now() - Duration::days(UNVERIFIED_THRESHOLD_DAYS);
         let unreferenced_paths = self
             .dataset
             .object_store
             .read_dir_all(&self.dataset.base, Some(self.before))
             .await?
-            .try_filter_map(|path| future::ready(self.path_if_not_referenced(path, &removal_stats)))
+            .try_filter_map(|obj_meta| {
+                // If a file is new-ish then it might be part of an ongoing operation and so we only
+                // delete it if we can verify it is part of an old version.
+                let maybe_in_progress =
+                    !self.delete_unverified && obj_meta.last_modified >= verification_threshold;
+                let path_to_remove =
+                    self.path_if_not_referenced(obj_meta.location, maybe_in_progress, &inspection);
+                if path_to_remove.is_ok() && path_to_remove.as_ref().unwrap().is_some() {
```

Comment:
nit: this might be clearer using `matches!()`
```suggestion
                if matches!(path_to_remove, Ok(Some(_))) {
```